rust/candid/src/types/reference.rs

```diff
@@ -0,0 +1,96 @@
+//! Data structure for Candid value Func and Service
+
+use super::{CandidType, Function, Serializer, Type};
+use ic_types::Principal;
+use serde::de::{self, Deserialize, Visitor};
+use std::convert::TryFrom;
+use std::{fmt, io::Read};
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct Func {
+    pub principal: Principal,
+    pub method: String,
+}
+#[derive(PartialEq, Debug, Clone)]
+pub struct Service {
+    pub principal: Principal,
+}
+
+impl CandidType for Func {
+    fn _ty() -> Type {
+        Type::Func(Function {
+            modes: Vec::new(),
+            args: Vec::new(),
+            rets: Vec::new(),
+        })
+    }
+    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_function(self.principal.as_slice(), &self.method)
+    }
+}
+impl CandidType for Service {
+    fn _ty() -> Type {
+        Type::Service(Vec::new())
+    }
+    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_principal(self.principal.as_slice())
+    }
+}
+
+impl<'de> Deserialize<'de> for Func {
+    fn deserialize<D>(deserializer: D) -> Result<Func, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct FuncVisitor;
+        impl<'de> Visitor<'de> for FuncVisitor {
+            type Value = Func;
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("Func value")
+            }
+            fn visit_byte_buf<E: de::Error>(self, bytes: Vec<u8>) -> Result<Func, E> {
+                if bytes[0] != 5u8 {
```

Comment:
Maybe I am lost, but what does the value `5` mean here?

Reply:
It's an ugly hack just to make serde happy. It's not related to the spec at all.

serde generates values via this `Visitor` trait, and it cannot consult any internal state in the deserializer. So the data we pass to the visitor should be self-contained. This `5` tag is an internal representation of `Function` value. This tag is generally not needed, because we can build different visitors for different values, e.g. `FuncVisitor`, `IntVisitor`. But to decode into the untyped `IDLValue`, we need this tag to distinguish different values.